consensus/executor.go

```diff
@@ -14,66 +21,445 @@ type executor[R any] struct {
-		// 	go func(consensusExec policy.ExecutionInternal[R], execIdx int) {
-		// 		result := innerFn(consensusExec)
-		// 		isFinalResult := int(resultCount.Add(1)) == e.requiredParticipants
-		// 		isCancellable := e.IsAbortable(result.Result, result.Error)
-		// 		if (isFinalResult || isCancellable) && resultSent.CompareAndSwap(false, true) {
-		// 			resultChan <- &execResult{result, execIdx}
-		// 		}
-		// 	}(executions[execIdx], execIdx)
-
-		// 	var result *execResult
-		// 	if execIdx < e.requiredParticipants {
-		// 		// timer := time.NewTimer(e.delayFunc(exec))
-		// 		// select {
-		// 		// case <-timer.C:
-		// 		// case result = <-resultChan:
-		// 		// 	timer.Stop()
-		// 		// }
-		// 		result = <-resultChan
-		// 	} else {
-		// 		result = <-resultChan
-		// 	}
-
-		// 	// Return if parent execution is canceled
-		// 	if canceled, cancelResult := parentExecution.IsCanceledWithResult(); canceled {
-		// 		return cancelResult
-		// 	}
-
-		// 	// Return result and cancel any outstanding attempts
-		// 	if result != nil {
-		// 		for i, execution := range executions {
-		// 			if i != result.index && execution != nil {
-		// 				execution.Cancel(nil)
-		// 			}
-		// 		}
-		// 		return result.result
-		// 	}
-		// }
+type execResult[R any] struct {
+	result   R
+	err      error
+	index    int
+	upstream common.Upstream
+}
+
+// resultToHash converts a result to a string representation for comparison
+func (e *executor[R]) resultToHash(result R, exec failsafe.Execution[R]) string {
+	return hashResult(e.resultToRawString(result, exec))
+}
+
+// resultToRawString converts a result to its raw string representation
+func (e *executor[R]) resultToRawString(result R, exec failsafe.Execution[R]) string {
+	if resp, ok := any(result).(*common.NormalizedResponse); ok {
+		if jr, err := resp.JsonRpcResponse(exec.Context()); err == nil {
+			return string(jr.Result)
+		}
+	}
+	return fmt.Sprintf("%+v", result)
+}
+
+// hashResult creates an MD5 hash of a string
+func hashResult(s string) string {
+	return fmt.Sprintf("%x", md5.Sum([]byte(s))) //#nosec G401 -- Permit usage of md5 for fast hashing, not used for security
+}
+
+func (e *executor[R]) Apply(innerFn func(failsafe.Execution[R]) *failsafeCommon.PolicyResult[R]) func(failsafe.Execution[R]) *failsafeCommon.PolicyResult[R] {
+	return func(exec failsafe.Execution[R]) *failsafeCommon.PolicyResult[R] {
+		parentExecution := exec.(policy.ExecutionInternal[R])
+		executions := make([]policy.ExecutionInternal[R], e.requiredParticipants)
+		executionWaitGroup := sync.WaitGroup{}
+		executionWaitGroup.Add(e.requiredParticipants)
+
+		// Create a buffered channel for the single result
+		resultChan := make(chan *failsafeCommon.PolicyResult[R], 1)
+
+		// Track responses for consensus checking
+		responses := make([]*execResult[R], 0, e.requiredParticipants)
+		responseMu := sync.Mutex{}
+
+		// Track used upstreams to ensure we don't use the same one twice
+		usedUpstreams := make(map[string]struct{})
```

Comment:
In terms of memory usage (think +10k RPS load) probably it'll be less to use struct as key:
```suggestion
		usedUpstreams := make(map[common.Upstream]struct{})
```
And something like:
```
// mark an upstream used
usedUpstreams[up] = struct{}{}

// check it
if _, seen := usedUpstreams[up]; seen {
    // we’ve already picked this one
}
```

Reply:
Added a small change to key by the memory address of the upstream as opposed to the string id.